compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/providers/impl/FirTypeResolverImpl.kt

```diff
@@ -142,15 +144,22 @@ class FirTypeResolverImpl(private val session: FirSession) : FirTypeResolver() {
             }
         }
 
-        for (scope in scopes) {
-            if (applicability == CandidateApplicability.RESOLVED) break
+        fun processClassifiers(scope: FirScope): Unit =
             scope.processClassifiersByNameWithSubstitution(qualifier.first().name) { symbol, substitutorFromScope ->
                 val resolvedSymbol = resolveSymbol(symbol, qualifier, qualifierResolver)
                     ?: return@processClassifiersByNameWithSubstitution
 
                 processCandidate(resolvedSymbol, substitutorFromScope)
             }
-        }
+
+        val selfTypeScope: FirScope? = scopes.find { scope -> scope is FirSelfTypeScope }
+        if (selfTypeScope != null && typeRef.render() == "Self")
```

Comment:
render() result should never be used for resolve purposes

Reply:
And in general, this logic looks very strange. Scopes should be just prioritized, without attempt of preselection. Please attempt to decide which priority `FirSelfTypeScope` should have.

Reply:
Removed, `FirSelfTypeScope` has maximum priority.